src/core/execute.c

```diff
@@ -419,33 +473,107 @@ int exec_spawn(Unit *unit,
         if (r < 0)
                 return log_unit_error_errno(unit, r, "Failed to serialize parameters: %m");
 
-        if (fseeko(f, 0, SEEK_SET) == (off_t) -1)
-                return log_unit_error_errno(unit, errno, "Failed to reseek on serialization stream: %m");
+        rewind(f);
 
-        r = fd_cloexec(fileno(f), false);
-        if (r < 0)
-                return log_unit_error_errno(unit, r, "Failed to set O_CLOEXEC on serialization fd: %m");
+        if (fds_array) {
+                r = send_many_fds(unit->manager->executors_pool_socket[0], fds_array, n_fds_array, 0);
+                if (r < 0)
+                        return log_unit_error_errno(unit, r, "Failed to send serialization fd: %m");
 
-        r = fdset_cloexec(fdset, false);
-        if (r < 0)
-                return log_unit_error_errno(unit, r, "Failed to set O_CLOEXEC on serialized fds: %m");
+                r = fd_wait_for_event(unit->manager->executors_pool_socket[0], POLLIN, 1 * USEC_PER_SEC);
```

Comment:
hmm, so you are calling poll() on a socket with O_NONBLOCK off. That's actually not OK. The problem is that sockets can drop packets anytime they like which means poll() is not reliable: it may report POLLIN without anything readable. if you then go and read something you might still hang.

I learnt that the hard way back in avahi days. Basically the rule is that poll() implies you are using O_NONBLOCK too.

Now, here I'd just do with both however. I'd simply stick to synchronous mode of the socket (i.e. leave O_NONBLOCK) but set SO_RCVTIMEO on it. This means any attempt to recv()/read()/recvmsg() will fail with a recognizable error, which we can then handle.